### PR TITLE
api-docs,catalog: Update messaging when no entities are in tables

### PR DESCRIPTION
**Pull request by adamdmharvey**
## Hey, I just made a Pull Request!

Minor change to the wording when there are no values in certain component tables. The intent is to help clarify in certain instances that it's not just that there's no info in the table (no provides API), but a bit more about what it means.

Was going to open an issue to discuss it, but figured easy enough to just file a PR and await any feedback/other wording suggestions, etc.  The primary for me was rewording "No component provides this API" to instead saying "This component does not provide it's own APIs".

I also made some of the kinds dynamic, which may not be necessary but figured some teams may have services but also libraries that might "provide" an API?

| Before | After |
| --- | --- |
|  image |  image |
| image | image |

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

File: plugins/catalog/src/components/HasComponentsCard/HasComponentsCard.tsx [hunk 1]
```diff
@@ -76,7 +76,7 @@ export const HasComponentsCard = ({ variant = 'gridItem' }: Props) => {
         <div>
           No component is part of this system.{' '}
           <Link to="https://backstage.io/docs/features/software-catalog/descriptor-format#kind-component">
-            Learn how to add components.
+            (Learn how to add components.)
```
Comment (Fox32) [inline plugins/catalog/src/components/HasComponentsCard/HasComponentsCard.tsx:79]: I think it somewhat looks strange with parentheses inside the link. And with a dot inside the parentheses.

File: plugins/api-docs/src/components/ApisCards/ConsumedApisCard.tsx [hunk 2]
```diff
@@ -72,9 +72,9 @@ export const ConsumedApisCard = ({ variant = 'gridItem' }: Props) => {
       variant={variant}
       emptyContent={
         <div>
-          No Component consumes this API.{' '}
+          This {entity.kind.toLowerCase()} does not consume any APIs.{' '}
           <Link to="https://backstage.io/docs/features/software-catalog/descriptor-format#specconsumesapis-optional">
-            Learn how to consume APIs.
+            (Learn how to change this.)
```
Comment (Fox32) [inline plugins/api-docs/src/components/ApisCards/ConsumedApisCard.tsx:77]: Would it be more consistent to write something like: "Learn how to consume APIs"?

File: plugins/api-docs/src/components/ApisCards/HasApisCard.tsx [hunk 3]
```diff
@@ -76,9 +76,9 @@ export const HasApisCard = ({ variant = 'gridItem' }: Props) => {
       variant={variant}
       emptyContent={
         <div>
-          No API is part of this system.{' '}
+          This {entity.kind.toLowerCase()} does not contain any APIs.{' '}
```
Comment (Fox32) [inline plugins/api-docs/src/components/ApisCards/HasApisCard.tsx:79]: I think whether we want to template this is a general design question for components in Backstage. We could make them very flexible and reusable in every case (with a lot of configuration and complexity) or make it easy do write custom variants of a component (more overhead by copy & paste). Right now the HasApisCard is just a thin layer around the entity table that contains the actual logic.

But I think for now it's the easiest route to go if you have such a usecase

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx [hunk 4]
```diff
@@ -74,7 +74,7 @@ describe('<ProvidedApisCard />', () => {
     );
 
     expect(getByText(/Provided APIs/i)).toBeInTheDocument();
-    expect(getByText(/No component provides this API/i)).toBeInTheDocument();
+    expect(getByText(/does not provide its own APIs/i)).toBeInTheDocument();
```
Comment (Fox32) [inline plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx:77]: "its own" vs "any" in the above consumed case, let's be consistent. I don't think that the API has to be owned by the component, you can also implement APIs provided by others.

File: plugins/api-docs/src/components/ApisCards/HasApisCard.tsx (hunk 3 above)
Comment (adamdmharvey) [reply plugins/api-docs/src/components/ApisCards/HasApisCard.tsx]: That's a good point.

* https://backstage.io/docs/features/software-catalog/system-model#docsNav
  * Here we say "A system, in this sense, is a collection of resources and components that exposes one or several public APIs..
  * So the verb is *exposes*.
* https://backstage.io/docs/features/software-catalog/descriptor-format#kind-system
  * Here we say, "The system may expose or consume one or several APIs."
  * So again "exposes".

Maybe exposes? Or maybe we should align back to provides to that doco here as well...

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx (hunk 4 above)
Comment (adamdmharvey) [reply plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx:77]: I think for me, if I _own_ a microservice and that service _provides_ an API, it's the API _for_ the microservice.  I might provide say only one API - my api - but I might consume 12 different APIs from other teams. Authentication APIs, user profile APIs, observability APIs, etc.  So I added "its" as possessive specifically to help enforce that it's an API fro\m this thing, from this component/this microservice, which is vastly different than just "it uses this other thing".

And then when you're in a system, a system might have 12 microservices, each one might then provide it's own APIs.  So the system as a whole may be a collection of 18 different APIs, but I interpret that the system has a collection of APIs, but it's not OWNED by the system. To take a real example, I might have a dozen squads own a dozen services, and each of them provide "their own" api, but then an architecture team or a product manager might own "the system" which has those APIs, and that's why I'm trying to veer away from saying "apis owned by a system" but rather APIs which are IN a system.

So I'll be honest, I really like this wording but I've not convinced you, have I? 😁

File: plugins/api-docs/src/components/ApisCards/HasApisCard.tsx (hunk 3 above)
Comment (Fox32) [reply plugins/api-docs/src/components/ApisCards/HasApisCard.tsx]: Exposing APIs from a system is a different functionality that isn't implemented. The idea is that an API is exposed for usage by other systems. PartOf is the relationship that is in use here.

Comment (adamdmharvey) [reply plugins/api-docs/src/components/ApisCards/HasApisCard.tsx]: I may have missed that nuance.  Are we saying that today in the Backstage model an API could be provided (`providesApi`), but not exposed (`partOf`), so the implication no one could actually call it from outside of itself?

`<joke/half serious>` Wouldn't that just be called a private function? `</joke/half serious>` 🤔

Comment (adamdmharvey) [reply plugins/api-docs/src/components/ApisCards/HasApisCard.tsx]: Or are we talking boundary of a system, like a microservice in a system could call another microservice API in the same system, but when you say expose you mean expose OUTSIDE of the boundary of a system?

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx (hunk 4 above)
Comment (Fox32) [reply plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx:77]: I think there are a lot of use cases where you provide an implementation for an API but are not owning it - I would even say it's the majority of cases. For example the format of a metric endpoint like prometheus. It's an API, and the API is defined by the consumer. Most of the time this isn't really visible and probably also not tracked in Backstage.

As it leads to confusion, I would explicitly avoid the term of "owning" here. I would reserve that for team ownership.

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx [hunk 5]
```diff
@@ -72,9 +72,9 @@ export const ProvidedApisCard = ({ variant = 'gridItem' }: Props) => {
       variant={variant}
       emptyContent={
         <div>
-          No component provides this API.{' '}
+          This {entity.kind.toLowerCase()} does not provide its own APIs.{' '}
           <Link to="https://backstage.io/docs/features/software-catalog/descriptor-format#specprovidesapis-optional">
-            Learn how to provide APIs.
+            (Learn how to change this.)
```
Comment (Fox32) [inline plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx:77]: Should we update this to the new style too?

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx (hunk 4 above)
Comment (adamdmharvey) [reply plugins/api-docs/src/components/ApisCards/ProvidedApisCard.test.tsx:77]: Ok.  Let me let my head clear a bit today, get to some other things and I'll take a fresh look, sort of step back and see if I understand at a broad stroke before taking another swag of wording.  Appreciate all the feedback!

I'll say I took a few of one of my team's existing Swagger definitions (defined statically in GitHub today along side static Markdown tables that link to them!)) and threw them in, and immediately these types of tables are really exposing what we have in a much more digestible way hence my wanting to dig into them further. So all great value add to an org to understand their engineering catalog.

File: plugins/api-docs/src/components/ApisCards/HasApisCard.tsx (hunk 3 above)
Comment (Fox32) [reply plugins/api-docs/src/components/ApisCards/HasApisCard.tsx]: So what is possible right now:

* Components can **provide** an API (which is a separate entity)
* Components can be **part of** a System.
* APIs (on their own) can be **part of** a System. For me that wouldn't mean that it is "exposed" or something like that, this is only a grouping.

Not possible:
* Systems can **expose/provide/make public** an API.

> OUTSIDE of the boundary of a system

Yes. It's about discoverability. Other teams responsible for other parts of a bigger picture might want to reuse existing things from other systems. But I think, at least personally, that an API that is reused by other people should have other "standards" especially when it comes to compatibility compared to team internal APIs.

As a small background about the initial wording: I chose to go very close to the wording of the relations. Beside that I later made a copy and paste mistake 😆 . I had an issue with the HasPart relationships as it wasn't that easy to add here compared to PartOf. That's the reason why I went with "contains".

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

**Review by Rugvip — APPROVED** [review]
Nice, thanks! :shipit:? :grin:

File: plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx [hunk 6]
```diff
@@ -71,11 +72,15 @@ export const ProvidedApisCard = ({ variant = 'gridItem' }: Props) => {
...
       variant={variant}
       emptyContent={
-        <div>
-          No component provides this API.{' '}
-          <Link to="https://backstage.io/docs/features/software-catalog/descriptor-format#specprovidesapis-optional">
-            Learn how to provide APIs.
-          </Link>
+        <div style={{ textAlign: 'center' }}>
+          <Typography variant="h6">
+            This {entity.kind.toLowerCase()} does not provide its own APIs.
```
Comment (Fox32) [inline plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx:77]: Actually is still think that we should remove the term "own" here and stay with "any"

Comment (Rugvip) [reply plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx:77]: Yep would agree with that

Comment (adamdmharvey) [reply plugins/api-docs/src/components/ApisCards/ProvidedApisCard.tsx]: Haven't really had a lot of time to look at this, so took your feedback here and applied it. Updated screen caps in PR description.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

**Comment by adamdmharvey** [comment]
oh wow; DCO fix is causing me all sorts of merging issues with the rest.json vs res.send challenge.  i seem to have broken my local fork and not sure how to fix it.  wah wahhhh.

```
backstage on  HEAD (bc60d2b) (REBASING 7/249) [=$✘»+?] via ⬢ v12.16.2 took 2s
❯ git status
HEAD detached from c41604bd5
You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Changes to be committed:
```

I may have to abandon this and start a new one.  (i wonder if DCO merging for all open, and older PRs, will also become challenging...)

**Comment by adamdmharvey** [comment]
Yeah super weird, it's triggered tons of merge conflicts which I didn't have before without signing, unless I somehow signed someone ELSES commits.

```
backstage on  HEAD (bc60d2b) (REBASING 7/249) [$?] via ⬢ v12.16.2
❯ git pull origin api-terminology
From github.com:adamdmharvey/backstage
 * branch                api-terminology -> FETCH_HEAD
Auto-merging plugins/techdocs-backend/src/service/router.ts
Auto-merging plugins/scaffolder-backend/src/service/router.ts
Auto-merging plugins/kubernetes-backend/src/service/router.ts
Auto-merging plugins/auth-backend/src/providers/aws-alb/provider.ts
CONFLICT (content): Merge conflict in plugins/auth-backend/src/providers/aws-alb/provider.ts
Auto-merging plugins/auth-backend/src/providers/aws-alb/provider.test.ts
CONFLICT (content): Merge conflict in plugins/auth-backend/src/providers/aws-alb/provider.test.ts
Auto-merging plugins/auth-backend/src/lib/oauth/OAuthAdapter.ts
CONFLICT (content): Merge conflict in plugins/auth-backend/src/lib/oauth/OAuthAdapter.ts
Auto-merging plugins/auth-backend/src/lib/oauth/OAuthAdapter.test.ts
CONFLICT (content): Merge conflict in plugins/auth-backend/src/lib/oauth/OAuthAdapter.test.ts
Auto-merging packages/techdocs-common/src/stages/publish/awsS3.ts
CONFLICT (content): Merge conflict in packages/techdocs-common/src/stages/publish/awsS3.ts
Auto-merging packages/config-loader/src/lib/schema/compile.ts
CONFLICT (content): Merge conflict in packages/config-loader/src/lib/schema/compile.ts
Auto-merging packages/cli/src/commands/index.ts
CONFLICT (content): Merge conflict in packages/cli/src/commands/index.ts
Auto-merging packages/cli/src/commands/config/schema.ts
CONFLICT (add/add): Merge conflict in packages/cli/src/commands/config/schema.ts
Auto-merging packages/backend-common/src/middleware/errorHandler.test.ts
Auto-merging microsite/sidebars.json
Auto-merging microsite/scripts/verify-sidebars.js
CONFLICT (add/add): Merge conflict in microsite/scripts/verify-sidebars.js
Removing docs/features/software-templates/extending/index.md
Removing docs/features/software-templates/extending/create-your-own-templater.md
```

**Comment by adamdmharvey** [comment]
Yeah that's def the problem:

```
commit e5b5ebe13445c59a26969ad63534101be470c96c (list)
Author: Andreas Stenius <git@astekk.se>
Date:   Fri Feb 19 14:43:25 2021 +0100

    config(schema): export mergeConfigSchemas from config-loader for use by backstage-cli config:schema.

    Signed-off-by: Andreas Stenius <git@astekk.se>
    Signed-off-by: Adam Harvey <adam.harvey@dxc.com>

commit e6ce7e2200994a513f470a8ce3c39f5c7ec2009e
Author: Andreas Stenius <git@astekk.se>
Date:   Fri Feb 19 11:17:15 2021 +0100

    docs(cli): describe config:schema

    Signed-off-by: Andreas Stenius <git@astekk.se>
    Signed-off-by: Adam Harvey <adam.harvey@dxc.com>
```

So when using the DCO to try to historically sign commits, it can't be done just based on pure history if any merge updates to rebase have done, and can in those scenarios might need to be cherry picked. (though i don't know HOW to do that yet)

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

**Comment by adamdmharvey** [comment]
Ok, phew. I think that's clean. I know @Rugvip you approved, but just asking you and @Fox32 for a final last check and see what you think on the wording, formatting, before this goes forward. Didn't get the time I really wanted to step back and look at the wording holistically, but have applied your feedback and it's an incremental which could continue to be adjusted!

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4640-mphxb0lxcs1ooqqgrrmysc3mjuaaze2n.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6034790dfcb22d9e035ce553

**Review by Rugvip — APPROVED** [review]
:shipit: 😁

**Comment by adamdmharvey** [comment]
ok phew, one core maintainer approval, a green pipeline, DCO is valid! shipping and we can iterate.

thanks for all the patience guys!!

**Pull request merged**